### PR TITLE
Add support for a CPU-only Mode

### LINKED ISSUE #1646: [FEA]: ControlMessage.tensors(TensorMemory) should support casting from the Python impl of TensorMemory
### Is this a new feature, an improvement, or a change to existing functionality?

Improvement

### How would you describe the priority of this feature request

High

### Please provide a clear description of problem this feature solves

Currently interacting with the ControlMessage requires `from morpheus._lib.messages import TensorMemory as CppTensorMemory`

### Describe your ideal solution

Support a cast in the python bindings

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I have searched the [open feature requests](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22%2Cimprovement%2Cenhancement) and have found no duplicates for this feature request

### LINKED ISSUE #1846: [EPIC] CPU Only Execution Mode
Tracking issue for the CPU only execution mode effort
```[tasklist]
### Tasks
- [ ] https://github.com/nv-morpheus/Morpheus/issues/562
- [x] Add API for getting/setting execution mode
- [x] Update sources and sinks to support CPU/GPU modes
- [x] Add suite of tests to ensure correct execution mode is being run
```

### LINKED ISSUE #1852: [BUG]: Copy constructor for `ControlMessage`  does not copy the payload, type or tensors
### Version

24.10

### Which installation method(s) does this occur on?

Source

### Describe the bug.

Current method only copies the tasks and config.

https://github.com/nv-morpheus/Morpheus/blob/branch-24.10/python/morpheus/morpheus/_lib/src/messages/control.cpp#L54

### Minimum reproducible example

_No response_

### Relevant log output

<details><summary>Click here to see error details</summary><pre>

 [Paste the error here, it will be hidden by default]

</pre></details>

### Full env printout

<details><summary>Click here to see environment details</summary><pre>

 [Paste the results of print_env.sh here, it will be hidden by default]

</pre></details>

### Other/Misc.

_No response_

### Code of Conduct

- [X] I agree to follow Morpheus' Code of Conduct
- [X] I have searched the [open bugs](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3Abug) and have found no duplicates for this bug report

### LINKED ISSUE #1955: [BUG]: Python impl of `TensorMemory.set_tensor` reshapes incoming tensors
### Version

24.10

### Which installation method(s) does this occur on?

_No response_

### Describe the bug.

The Python impl of `TensorMemory.set_tensor` will reshape the incoming tensor to 2 dimensions. 

The problem is the C++ impl doesn't perform this, neither does the Python impl of the `TensorMemory` constructor or `TensorMemory.set_tensors`.

### Minimum reproducible example

```python
import cupy as cp

from morpheus.config import CppConfig
from morpheus.messages.memory.tensor_memory import TensorMemory

def repo(use_cpp:bool):
    CppConfig.set_should_use_cpp(use_cpp)
    mem = TensorMemory(count=3, tensors={"input_ids": cp.array([1, 2, 3]), "input_mask": cp.array([1, 1, 1]), "segment_ids": cp.array([0, 0, 1])})
    mem.set_tensor('new_tensor', cp.array([4, 5, 6]))
    print(f"use_cpp: {use_cpp}\n{mem.get_tensors()}\n")


repo(True)
repo(False)
```

### Relevant log output

<details><summary>Click here to see error details</summary><pre>

use_cpp: True
{'input_ids': array([1, 2, 3]), 'input_mask': array([1, 1, 1]), 'new_tensor': array([4, 5, 6]), 'segment_ids': array([0, 0, 1])}

use_cpp: False
{'input_ids': array([1, 2, 3]), 'input_mask': array([1, 1, 1]), 'segment_ids': array([0, 0, 1]), 'new_tensor': array([[4],
       [5],
       [6]])}

</pre></details>

### Full env printout

<details><summary>Click here to see environment details</summary><pre>

 [Paste the results of print_env.sh here, it will be hidden by default]

</pre></details>

### Other/Misc.

_No response_

### Code of Conduct

- [x] I agree to follow Morpheus' Code of Conduct
- [x] I have searched the [open bugs](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3Abug) and have found no duplicates for this bug report

### PR DESCRIPTION
**Pull request by dagardner-nv**
## Description
* Adds a new enum `morpheus.config.ExecutionMode` with members `GPU` & `CPU` along with a new `morpheus.config.Config.execution_mode` attribute.
* For backwards compatibility, by default `Config.execution_mode` will always default to `GPU`
* Add new `supported_execution_modes` to `StageBase` which returns `ExecutionMode.GPU` by default. This ensures that building a pipeline with a stage not matching the execution mode will raise a reasonable error to the user.
* Add `CpuOnlyMixin` and `GpuAndCpuMixin` mixins to automate overriding this, and makes it easier for users to determine which execution modes a given stage supports at a glance.
* Since C++ Stage/Message impls can only support cuDF DataFrames, and RMM tensors, this PR re-purposes the existing Python stage/message impls mode to serve as CPU-only mode.
* CPU-only mode will center around pandas DataFrames and NumPy arrays for tensors, since the current Python code which expects cuDF/CuPy is already 99% compatible with pandas/NumPy.
* Avoid importing `cudf` or any other GPU based package which will fail on import at the top-level of a module. This is important for stage, message and modules which are automatically imported by the morpheus CLI tool.
* Add new utility methods to `morpheus.utils.type_utils` (ex: `get_df_pkg`, `is_cudf_type`) to help avoid importing cudf directly
* Add a new `Config.freeze` method which will make a config object immutable. This will be called the first time a config object is used to construct a pipeline or stage object. Prevents the possibility of config parameters from being changed in the middle of pipeline construction.
* `CudfHelper::load` is no longer called automatically on import, instead it is called manually on pipeline build when execution mode is GPU.
* Add Python implementation of `ControlMessage`
* To simulate a system without a GPU to test CPU-only mode, if the `CPU_ONLY` environment variable is defined `docker/run_container_dev.sh` will launch the container using the `runc` runtime.
* Remove automatic test parameterization of C++/Python mode, since supporting CPU-only mode will become the exception not the rule. Add a new `gpu_and_cpu_mode` test marker to explicitly indicate a test intended to be parameterized over execution modes.
* Fix copy constructor for `ControlMessage`
* `AppShieldSourceStage` now emits `ControlMessage`s, `AppShieldMessageMeta` is now deprecated
* `AutoencoderSourceStage` and thus `AzureSourceStage`, `CloudTrailSourceStage`, and `DuoSourceStage` now emit `ControlMessage`, `UserMessageMeta` is now deprecated.
* DFP production pipeline updated to remove `DFPMessageMeta`, pipeline now executes in C++ mode.
* Consolidate common logig in `docker/run_container_dev.sh` & `docker/run_container_release.sh` into `docker/run_container.sh`
* Remove inconsistent behavior in the Python impl of `TensorMemory.set_tensor` (#1955)

Closes #1646
Closes #1846
Closes #1852
Closes #1955

## By Submitting this PR I confirm:
- I am familiar with the [Contributing Guidelines](https://github.com/nv-morpheus/Morpheus/blob/main/docs/source/developer_guide/contributing.md).
- When the PR is ready for review, new or existing tests cover these changes.
- When the PR is ready for review, the documentation is up to date with these changes.

### CODE REVIEW

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
Dont have too many comments. Just need to figure out the LLM part before we move forward.

File: python/morpheus/morpheus/cli/commands.py [hunk 1]
```diff
@@ -296,8 +302,19 @@ def install(**kwargs):
...
+        logger.warning("The --use_cpp flag is deprecated and will be removed in a future release")
+
     # Since the option isnt the same name as `should_use_cpp` anymore, manually set the value here.
-    CppConfig.set_should_use_cpp(kwargs.pop("use_cpp", CppConfig.get_should_use_cpp()))
+    use_cpu_only = kwargs.pop("use_cpu_only")
+    use_cpp = kwargs.pop("use_cpp")
+    if use_cpu_only:
+        CppConfig.set_should_use_cpp(False)
+    else:
+        CppConfig.set_should_use_cpp(use_cpp)
```
[the remark this answers is not in the document]
Comment (mdemoret-nv) [reply python/morpheus/morpheus/cli/commands.py:314]: Lets add a comment here.

File: docker/run_container_dev.sh [hunk 2]
```diff
@@ -28,7 +28,15 @@ DOCKER_IMAGE_NAME=${DOCKER_IMAGE_NAME:-"morpheus"}
...
-DOCKER_ARGS="--runtime=nvidia --env WORKSPACE_VOLUME=${PWD} -v $PWD:/workspace --net=host --gpus=all --cap-add=sys_nice"
+DOCKER_ARGS=" --env WORKSPACE_VOLUME=${PWD} -v $PWD:/workspace --net=host --cap-add=sys_nice"
+
+if [[ -n "${CPU_ONLY}" ]]; then
+   echo -e "${b}Executing in CPU only mode${x}"
+   DOCKER_ARGS="${DOCKER_ARGS} --runtime=runc"
+else
+    echo -e "${b}Executing in GPU mode${x}"
+    DOCKER_ARGS="${DOCKER_ARGS} --runtime=nvidia --gpus=all"
+fi
```
[the remark this answers is not in the document]
Comment (dagardner-nv) [reply docker/run_container_dev.sh:39]: Currently `run_container_dev.sh` and `run_container_release.sh` don't share a common script the way the `build_*.sh` do.

Let me know if you want me to try and address that in this PR.

File: python/morpheus/morpheus/stages/input/file_source_stage.py [hunk 3]
```diff
@@ -99,6 +99,8 @@ def __init__(self,
         self._iterative = iterative
         self._repeat_count = repeat
 
+        self._df_type = exec_mode_to_df_type_str(c.execution_mode)
+
```
Comment (mdemoret-nv) [inline python/morpheus/morpheus/stages/input/file_source_stage.py:103]: This pattern is happening a lot with classes that derive from `CpuAndGpuMixin`. We should make this a function on the mixin or have the mixin set this property. We can replicate this with any function that uses `Config.execution_mode` and is used by to `CpuAndGpuMixin`

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
Couple more suggestions

File: tests/conftest.py [hunk 4]
```diff
@@ -213,42 +194,6 @@ def pytest_runtest_teardown(item, nextitem):
...
-    yield do_use_cpp
-
-
-# This fixture will be used by all tests.
-@pytest.fixture(scope="function")
-def use_cpp(_set_use_cpp: bool):
-
-    # Just return the set value
-    yield _set_use_cpp
-
```
Comment (mdemoret-nv) [inline tests/conftest.py:250]: Below this, the fixtures `config_only_cpp` and `config_no_cpp` were unchanged. Should these be updated too?

File: tests/conftest.py [hunk 5]
```diff
@@ -295,7 +243,7 @@ def df_type(request: pytest.FixtureRequest):
         use_cudf = request.node.get_closest_marker("use_cudf") is not None
 
         if (use_pandas and use_cudf):
-            raise RuntimeError(f"Both markers (use_cpp and use_python) were added to function {request.node.nodeid}. "
+            raise RuntimeError(f"Both markers (gpu_mode and cpu_mode) were added to function {request.node.nodeid}. "
```
Comment (mdemoret-nv) [inline tests/conftest.py:246]: I think this is an error. Shouldnt it be `use_pandas` and `use_cudf`? The original was wrong too

File: tests/conftest.py [hunk 6]
```diff
@@ -304,22 +252,72 @@ def df_type(request: pytest.FixtureRequest):
...
+        gpu_mode = request.node.get_closest_marker("gpu_mode") is not None
+        cpu_mode = request.node.get_closest_marker("cpu_mode") is not None
+
+        if (gpu_mode and cpu_mode):
+            raise RuntimeError(f"Both markers (gpu_mode and cpu_mode) were added to function {request.node.nodeid}. "
+                               "Use the gpu_and_cpu_mode marker to test both.")
+
+        # if both are undefined, infer based on the df_type
+        if (not gpu_mode and not cpu_mode):
+            gpu_mode = request.node.get_closest_marker("use_cudf") is not None
```
Comment (mdemoret-nv) [inline tests/conftest.py:272]: This line is never used. I think if we are going to infer the results, we should only infer if there is one or the other. Not both.

File: python/morpheus/morpheus/io/deserializers.py [hunk 7]
```diff
@@ -60,25 +61,27 @@ def _read_file_to_df_py(*,
...
+        reader = get_json_reader(df_type)
+        df = reader(file_name, **kwargs)
+
+    else:
+        df_class = df_type_str_to_pkg(df_type)
 
-    elif (mode == FileTypes.CSV):
-        df: pd.DataFrame = df_class.read_csv(file_name, **kwargs)
+        if (mode == FileTypes.CSV):
+            df: DataFrameType = df_class.read_csv(file_name, **kwargs)
```
[the remark this answers is not in the document]
Comment (dagardner-nv) [reply python/morpheus/morpheus/io/deserializers.py:73]: The primary reason for this is that when cudf is needed, the ability to return `functools.partial(cudf.read_json, engine='cudf')` while CSV & PARQUET need no such changes.

File: tests/conftest.py (hunk 4 above)
Comment (dagardner-nv) [reply tests/conftest.py:249]: good catch

File: tests/conftest.py (hunk 5 above)
Comment (dagardner-nv) [reply tests/conftest.py:212]: Good catch, I think I just did a find/replace here

File: tests/conftest.py (hunk 6 above)
Comment (dagardner-nv) [reply tests/conftest.py:238]: It looks like you're right inferring gpu mode based on the `use_cudf` marker is not used. On the other hand, a large number of tests are relying upon the `use_pandas` marker for setting python/cpu mode.

**Review by mdemoret-nv — APPROVED** [review]

**Pull request merged**